core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,392 @@
+    }
+    if (mPartitionOffset > UPLOAD_THRESHOLD) {
+      uploadPart();
+    }
+    waitForAllPartsUpload();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    // Set the closed flag, we never retry close() even if exception occurs
+    mClosed = true;
+
+    // Multi-part upload has not been initialized
+    if (!mMultiPartUploadInitialized) {
+      if (mFile == null) {
+        LOG.debug("Streaming upload output stream closed without uploading any data.");
+        RetryUtils.retry("put empty object for key" + mKey, () -> createEmptyObject(mKey),
+            mRetryPolicy);
+      } else {
+        try {
+          mLocalOutputStream.close();
+          final String md5 = mHash != null ? Base64.encodeBase64String(mHash.digest()) : null;
+          RetryUtils.retry("put object for key" + mKey, () -> putObject(mKey, mFile, md5),
+              mRetryPolicy);
+        } finally {
+          if (!mFile.delete()) {
+            LOG.error("Failed to delete temporary file @ {}", mFile.getPath());
+          }
+        }
+      }
+      return;
+    }
+
+    try {
+      if (mFile != null) {
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
+        uploadPart(mFile, partNumber, true);
+      }
+
+      waitForAllPartsUpload();
+      RetryUtils.retry("complete multipart upload",
+          this::completeMultiPartUploadInternal, mRetryPolicy);
+    } catch (Exception e) {
+      LOG.error("Failed to upload {}", mKey, e);
+      throw new IOException(e);
+    }
+  }
+
+  /**
+   * Creates a new temp file to write to.
+   */
+  private void initNewFile() throws IOException {
+    mFile = new File(PathUtils.concatPath(CommonUtils.getTmpDir(mTmpDirs), UUID.randomUUID()));
+    initHash();
+    if (mHash != null) {
+      mLocalOutputStream =
+          new BufferedOutputStream(new DigestOutputStream(new FileOutputStream(mFile), mHash));
+    } else {
+      mLocalOutputStream = new BufferedOutputStream(new FileOutputStream(mFile));
+    }
+    mPartitionOffset = 0;
+    LOG.debug("Init new temp file @ {}", mFile.getPath());
+  }
+
+  private void initHash() {
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
+      mHash = null;
+    }
+  }
+
+  /**
+   * Uploads part async.
+   */
+  protected void uploadPart() throws IOException {
+    if (mFile == null) {
+      return;
+    }
+    if (!mMultiPartUploadInitialized) {
+      RetryUtils.retry("init multipart upload", this::initMultiPartUploadInternal, mRetryPolicy);
+      mMultiPartUploadInitialized = true;
+    }
+    mLocalOutputStream.close();
+    int partNumber = mPartNumber.getAndIncrement();
+    File newFileToUpload = new File(mFile.getPath());
+    mFile = null;
+    mLocalOutputStream = null;
+    uploadPart(newFileToUpload, partNumber, false);
```

Comment:
can inline all the parameter
uploadPart(new File(mFile.getPath()), mPartNumber.getAndIncrement(), false);